Assets/Scripts/Calculator.cs

```diff
@@ -0,0 +1,64 @@
+// Copyright 2021 The Aha001 Team.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Text;
+
+// The calculator that executes input expressions.
+public class Calculator {
+  private const int _maxChars = 100;
+  private readonly CalculatorScreen _screen = null;
+  private readonly StringBuilder _cachedInput = new StringBuilder();
+
+  private bool _hasError = false;
+
+  public bool HasError => _hasError;
+
+  public Calculator(CalculatorScreen screen) {
+    _screen = screen;
+  }
+
+  public void OnInput(string input) {
+    // If the calculator is in an error state, only the "AC" button can reset it.
+    if (input == CalculatorInput.AllClear) {
+      _cachedInput.Clear();
+      _screen.Clear();
+      _hasError = false;
+    }
+    if (HasError) {
+      return;
+    }
+
+    if (CalculatorInput.TryGetPrintable(input, out string printable)) {
+      _cachedInput.Append(printable);
+      if (_cachedInput.Length > _maxChars) {
+        ReportError(CalculatorScreen.Error.Overflow);
+      } else {
+        // TODO: invoke the underlying SeedLang engine to validate the cached expression.
+        _screen.Print(_cachedInput.ToString());
```

Comment:
Just thinking about the calculator model needs notify the screen to refresh in several places. Is there any kind of MVVM framework in Unity3D to bind views and models to make the notification simpler. Do you think if we need this kind of framework when the UI is more complex?

Reply:
Yes, I agree that an MVVM/MVC pattern will be required once the logic becomes more complex.

There are a few 3rd-party MVC libraries but it seems none of them is popular and actively maintained enough. A typical way to implement MVVM in Unity is to leverage the event-dispatcher-handler pattern, e.g., to use System.EventHandler delegate to connect the model class and the corresponding game objects.

The calculator app has a rather simple logic as for the model/view relationship. I will keep this in mind if the relationship turns to be complex in the future.